src/main/java/gregtech/api/capability/impl/FuelRecipeLogic.java

```diff
@@ -207,21 +224,88 @@ public long getMaxVoltage() {
+    }
+
+    public int getFuelAmount() {
+        return calculateFuelAmount(this.previousRecipe);
+    }
+
+    protected double calculateRecipeDurationMultiplier() {
+        return 1.0;
     }
 
     protected int calculateRecipeDuration(FuelRecipe currentRecipe) {
-        return currentRecipe.getDuration();
+        return (int) (currentRecipe.getDuration() * calculateRecipeDurationMultiplier());
+    }
+
+    public int getRecipeDuration() {
+        return calculateRecipeDuration(this.previousRecipe);
+    }
+
+    /**
+     * this method aim to tell if the current recipe has ended in order to check for a new recipe see {@link #isActive}
+     * or {@link #isWorkingEnabled} if you want to know if the machine is running.
+     *
+     * @return true if no recipe is running false otherwise
+     */
+    public boolean hasRecipeEnded() {
+        return this.recipeDurationLeft <= 0;
+    }
+
+    public boolean canProduceEnergy() {
+        return !hasRecipeEnded();
+    }
+
+    public boolean canConsumeFuel() {
+        return !hasRecipeEnded();
+    }
+
+    /**
+     * Compute a multiplier used by default in {@link #startRecipe(FuelRecipe, int, int)}  startRecipe} each time the recipe start.
+     * fast changing efficiency could be implemented in {@link #calculateDynamicEnergyEfficiency() calculateDynamicEnergyEfficiency} for better responsiveness of the changes.
+     *
+     * @return power multiplier
+     */
+    protected double calculateStaticEnergyEfficiency() {
+        return 1.0;
+    }
+
+    /**
+     * Compute a multiplier used default in {@link #calculateRecipeOutputVoltage() calculateRecipeOutputVoltage} each tick.
+     * constant/rarely changing efficiency should/could be implemented in {@link #calculateStaticEnergyEfficiency() calculateStaticEnergyEfficiency}
+     *
+     * @return power multiplier
+     */
+
+    protected double calculateDynamicEnergyEfficiency() {
+        return 1.0;
+    }
+
+    /**
+     * Apply {@link #calculateDynamicEnergyEfficiency() calculateDynamicEnergyEfficiency} multiplier
+     * to the outputVoltage returned by {@link #getRecipeOutputVoltage(), getRecipeOutputVoltage}
+     * in order to get the effective power output used in the {@link #update() update} method.
+     *
+     * @return Effective power output of the machine.
+     */
+    public long calculateRecipeOutputVoltage() {
+        return (long) (this.recipeOutputVoltage * calculateDynamicEnergyEfficiency());
     }
 
     /**
      * Performs preparations for starting given recipe and determines it's output voltage
+     * the {@link #calculateStaticEnergyEfficiency() calculateStaticEnergyEfficiency} multiplier is by default applied in this method.
      *
      * @return recipe's output voltage
      */
     protected long startRecipe(FuelRecipe currentRecipe, int fuelAmountUsed, int recipeDuration) {
```

Comment:
What's even the point of this method returning any value if you have it recomputed every tick?

Reply:
well let's say backward compatibility ? no because you see in #1604 I ignored the value returned at one time and a wise guy told me it wasn't a good idea

Reply:
I repeat, backwards compatibility is really no an issue, especially for that particular method